### ISSUE TITLE
PlayerNotificationManager - Improve APIs for setting player notification priority

### ISSUE THREAD

**Issue by hobsRKM** [google/ExoPlayer#7820]
**Note: Works fine on lower API < 29**

On android 10 api 29
I tried the exoplayer example form codelabs and implemented player notifications.

The notification icon in status bar is not shown, but displays in notification drawer as **other** categories with low importance.

Tried setting it to priority high still doesn't show up in Top status bar.

Here's the example code from google.

**playerNotificationManager.setPriority(NotificationCompat.PRIORITY_MAX**);

```
**player.setPlayWhenReady(playWhenReady);
    player.seekTo(currentWindow, playbackPosition);
    player.addListener(playbackStateListener);
    player.prepare(mediaSource, false, false);
    ..............................
 
    **playerNotificationManager.setPlayer(player);
    playerNotificationManager.setColorized(true);
    playerNotificationManager.setPriority(NotificationCompat.PRIORITY_MAX);
    playerNotificationManager.setColor(getResources().getColor(R.color.colorAccent));**

    //mediaSessionConnector.setPlayer(player, null, customAction1.class);

  }
.....................
  private void releasePlayer() {
    if (player != null) {
      playbackPosition = player.getCurrentPosition();
      currentWindow = player.getCurrentWindowIndex();
      playWhenReady = player.getPlayWhenReady();
      player.removeListener(playbackStateListener);
      player.release();
      player = null;
    }
  }

  private MediaSource buildMediaSource(Uri uri) {
    DataSource.Factory dataSourceFactory =
            new DefaultDataSourceFactory(this, "exoplayer-codelab");
    DashMediaSource.Factory mediaSourceFactory = new DashMediaSource.Factory(dataSourceFactory);
    MediaSource mediaSource1 = mediaSourceFactory.createMediaSource(uri);

    // Additionally, create a media source using an MP3.
    MediaSource mediaSource2 = mediaSourceFactory.createMediaSource(uri);

    return new ConcatenatingMediaSource(mediaSource1, mediaSource2);
  }

  @SuppressLint("InlinedApi")
  private void hideSystemUi() {
    playerView.setSystemUiVisibility(View.SYSTEM_UI_FLAG_LOW_PROFILE
            | View.SYSTEM_UI_FLAG_FULLSCREEN
            | View.SYSTEM_UI_FLAG_LAYOUT_STABLE
            | View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY
            | View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
            | View.SYSTEM_UI_FLAG_HIDE_NAVIGATION);
  }

  private class PlaybackStateListener implements Player.EventListener{

    @Override
    public void onPlayerStateChanged(boolean playWhenReady,
                                     int playbackState) {
      String stateString;
      switch (playbackState) {
        case ExoPlayer.STATE_IDLE:
          stateString = "ExoPlayer.STATE_IDLE      -";
          break;
        case ExoPlayer.STATE_BUFFERING:
          stateString = "ExoPlayer.STATE_BUFFERING -";
          break;
        case ExoPlayer.STATE_READY:
          stateString = "ExoPlayer.STATE_READY     -";
          break;
        case ExoPlayer.STATE_ENDED:
          stateString = "ExoPlayer.STATE_ENDED     -";
          break;
        default:
          stateString = "UNKNOWN_STATE             -";
          break;
      }
      Log.d(TAG, "changed state to " + stateString
              + " playWhenReady: " + playWhenReady);
    }
  }

}**

```

**Comment by marcbaechinger** [comment]
The value set with [setPriority(@Priority int priority)](https://github.com/google/ExoPlayer/blob/r2.11.8/library/ui/src/main/java/com/google/android/exoplayer2/ui/PlayerNotificationManager.java#L888) has only effect on API 25 and lower. 

For higher API levels, the importance of the notification is determined by the [importance level of notification channel](https://developer.android.com/training/notify-user/channels#importance) to which the notification belongs to. By default this is [IMPORTANCE_LOW](https://github.com/google/ExoPlayer/blob/r2.11.8/library/ui/src/main/java/com/google/android/exoplayer2/ui/PlayerNotificationManager.java#L455). If you want to set the importance to another value, you can create your own notification channel and [pass the channel id to the constructor of the `PlayerNotificationManager`](https://github.com/google/ExoPlayer/blob/r2.11.8/library/ui/src/main/java/com/google/android/exoplayer2/ui/PlayerNotificationManager.java#L548).

I marked this issue as a enhancement and documentation candidate. At least the java doc of `setPriority` should somehow give a better advice on what to do for API levels above 25.

**Comment by hobsRKM** [comment]
I tried on an emulator on API 29 works perfect, But not sure why on the real device with Android 10 it goes to "Other" Category instead of important and hides the status icon. (device tested one plus 6t)

**Comment by marcbaechinger** [comment]
Are you using your own notification channel?

**Comment by hobsRKM** [comment]
Yes.
I created a channel with id 987 to test and passed the same id to PlayerNotifcation constructor.

Please let me know if it's correct.
```

public void createChannel() {

        // create android channel
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            NotificationChannel androidChannel = new NotificationChannel("987",
                    "test_channel", NotificationManager.IMPORTANCE_HIGH);
            // Sets whether notifications posted to this channel should display notification lights

            getManager().createNotificationChannel(androidChannel);


        }

    }
  private NotificationManager getManager() {
        if (mManager == null) {
            mManager = (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);
        }
        return mManager;
    }

```

**Comment by marcbaechinger** [comment]
This seems to be correct yes.

Please allow me to make you aware that making media notifications of `IMPORTANCE_HIGH` is not recommended. The `PlayerNotificationManager` posts updates the notification quite often, for instance when a player status change happens. This causes the notification to be reposted for example when the user pauses playback, the player starts to buffer, a new song starts or for various other reasons. 

This is the reason why the manager by default uses `IMPORTANCE_LOW`. You can set your own importance but please be aware that the system then does play a sound and shows a heads up notification when the notification is updated (you can not change that system behaviour). As a user who wants to listen to audio this is highly disturbing I would say and once you get it working with `IMPORTANCE_HIGH` I think you will see that this is probably not the right thing to do for the users sake.

I just made it working with the code below (it's the same as you have) and for me as a user it is unacceptable that the notification always beeps when updated (eg. when I press pause/play) or a heads up notification pops up every time the notification gets updated (see also #7806).

```
private void createNotificationChannel() {
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
      int importance = NotificationManager.IMPORTANCE_HIGH;
      NotificationChannel channel =
          new NotificationChannel(CHANNEL_ID, getString(R.string.channel_name), importance);
      channel.setDescription(getString(R.string.channel_description));
      NotificationManager notificationManager = getSystemService(NotificationManager.class);
      notificationManager.createNotificationChannel(channel);
    }
  }
```

Another note: If you have run your app already and the notification channel has been created already you may want to deinstall the app completely before redeploying. This makes sure that any notification channels which have been created with this app and potentially the same channel id are removed and your new setting takes effect. After playing around with this It appears to me that some settings in the "Apps and notification" of the Android settings are preserved. So best is to manually deinstall and then deploy from Android Studio again.

**Comment by hobsRKM** [comment]
OK thanks, will look into it.

It would be good if there was an additional argument to player notifications constructor to override the importance. Rather than creating a new channel.

**Comment by marcbaechinger** [comment]
Agreed that making the importance an argument possibly would make it easier and we should probably think about this.

I don't think this additional argument needs to be on the constructor though. The static factory methods would be a better place. Said this, there are many arguments already and adding another would add another static method. So we should probably think of replacing these with a builder to avoid large list of parameters and many factory methods.

Another improvement would be that we could embrace fragmentation a bit by creating a new `NotificationChannel` with a corresponding `IMPORTANCE_XYZ` level when the `setPriority()` method is called. We would then, vice versa, need to adjust the priority, if a `NotificationChannel` with another importance level than LOW is used. This way we could keep both of these in sync with each other. 

I'm marking this as an enhancement. We won't work on this right away though and I can not give any timeline whether and when that would happen. (change regarding better java doc is done and landing here soon)
